### ISSUE TITLE
filter k8s objects by _anything_ in .status?

### ISSUE THREAD

**Issue by x3nb63** [jgriff/k8s-resource#13]
looking at filtering by phase I realize its not sufficient for what I wanna do: I have about 10 `ReplicationController` objects (where each may have many pods) .

Now I want at resource that triggers when _all 10_ have at least one replica in _Ready_ state (cause that state appears to respect eventual `readinessProbe` and `livenessProbe` definitions on the pods containers - `Running` phase on pods does not).

So I would need to

1. filter all ReplicationControllers by name with AND to ensure I get all 10 or nothing
2. inspect their .status.readyReplicas to be >= 1

on point 1) can filter `name` be an array? Since its only 10 I can get away with defining 10 resources, for my case at least, looks impressive on WebUI but is not exactly nice.

for point 2) a ReplicationController looks like this, no "phase" field.
```
$ kubectl get rc/mgr -o json | jq .status
{
  "availableReplicas": 1,
  "fullyLabeledReplicas": 1,
  "observedGeneration": 1,
  "readyReplicas": 1,
  "replicas": 1
}
```
which brings me to the actual question:

How about a general filter by status which takes a arbitrary string thats passed to the jq query part? 

ie. resulting in `kubectl get rc/mgr -o json | jq '.status.readyReplicas >= 1'`

This would allow to filter any object by anyting, I guess.

(I can try producing a pull request, but some guidance how you'd approach this would be nice)

**Comment by jgriff** [comment]
I like the idea of a general "all else fails, write your own" filter.  The current filter support provides first class concepts (`name`, `olderThan`, `phases`) in an effort to remove the cognitive burden for how the filter is applied (at a JSON resource level), and to simplify the resource configuration.

But I can see value in having the ability to drop down a level and write your own `jq` filter if you have more custom needs.


> (I can try producing a pull request, but some guidance how you'd approach this would be nice)

A pull request for this would be great!  Below are my thoughts...

## Resource Config

Something like this seems reasonable for the resource config:
```yaml
      filter:
        jq: [".status.readyReplicas >= 1"]
```

I think it should support a list, so you can chain multiple filters without having to write them all in one expression:
```yaml
      filter:
        jq: 
          - ".status.replicas >= 1"
          - ".status.replicas == .status.readyReplicas"
```
(just as an example)

## Implementation

This seems like a straightforward add of a `filterByJqExpressions()` function to the `check` script, modeled after the other filter functions([`filterByName()`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/check#L35), [`filterByCreationOlderThan()`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/check#L46), [`filterByPhases()`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/check#L57)).

Then invoke it after the last current filter function [`filterByPhases()`](https://github.com/jgriff/k8s-resource/blob/b2d1d5a0f0db57ead8427809449894a8388330a1/assets/check#L14)
```bash
main() {
    log "\n\n--[CHECK]-------------------------"
    extractPreviousVersion
    queryForVersions
    filterByName
    filterByCreationOlderThan
    filterByPhases
    filterByJqExpressions
    pareDownVersionInfo
    emitResult
}
```

Start with adding bats tests for the new filter in [`check.bats`](https://github.com/jgriff/k8s-resource/blob/master/test/check.bats), again modeled after the existing filter tests.  Try to also cover corner cases where the user misconfigured the jq expression, or gave you a bad filter.  Any "bad" filter expression should be an error from `check`.

**Comment by x3nb63** [comment]
here is my first attempt implementing `check` -> https://github.com/x3nb63/k8s-resource/commit/3d306c6fc449b05ad5ef6230ccf41e5189699c93

it works as far as `check.bats` covers but I am not entirely happy: I failed using `jq --argjson FILTER_QUERIES ...` to pass in the array of query strings. So the whole thing now relies on bash's lazy string handling ... requiring the query writer to properly escape the json ...

problem is that I could not figure how to make jq evaluate a passed in string and treat it as expression and not as string. It was always comparing the query string with element content rather than doing the query...  if you have an idea? ... Maybe its just not possible.

that aside I start working it into `in` and `out` and will prepare a pull request once I got use out of it. I have a fairly complex pipeline to write and shape it along the way...
